### ISSUE TITLE
OpenAPI support?

### ISSUE THREAD

**Issue by ianthetechie** [valhalla/valhalla#2960]
Do you have any interest in OpenAPI for documentation / (usually dumb, but usable) client generation? I'm working on a spec at the moment and need to document like 98% of the API (sans public transit). I'm not sure how widespread OpenAPI usage is, but I figured it might be of use to have a canonical version as part of this repo. In the future you may even be able to automate this, but having something to start with might help as the OpenAPI spec is a bit confusing.

If this isn't useful/needed with your existing process, no worries but I'd be happy to open a PR with my yaml file if you're interested.

**Comment by kevinkreiser** [comment]
@iandees in my opinion, if it didnt replace the existing technical API docs, then it would only make sense if its automated from the start. if it doesnt start out that way then it will be essentially equivalent to the api documentation we have today (which is various forms of out of date haha). if however we can replace the existing technical documentation with it in some for or another (i guess by generating markdown?) it seems reasonable to me even if its not automated yet because at least its not duplication. i suppose id need to go read a bit about OpenAPI documentation.

another question i have is just to make sure you are talking about the http json api? because internally we do have a protobuf api for all of the inputs and a decent chunk of the outputs (all the ones that produce a route or trace_attributes). i've been meaning to clean up the pbf a bit in preparation for gRPC support which i think is a natural extension of our design, pbf all the way through. maybe that would be a better course of action in the end because then you get the generated client and API defintionally although you dont get docs for free so far as i am aware.

**Comment by ianthetechie** [comment]
Hey Kevin, good points all around.

Regarding replacing the existing docs (https://valhalla.readthedocs.io/en/latest/), this could theoretically replace those in large part, though I have to say the doc generators aren't the greatest from a beauty perspective and the whole OpenAPI idea is focused on technical docs describing the API, the model/schema objects, what's required, the range of acceptable values, and the like. The existing docs fulfill a broader role. This is a question you and your team would have to answer, but one possible route would be splitting API docs (what does this do/what's the "grammar" I need to know, the meaning of each enum, etc.) from more general prose technical docs (how does this all work conceptually).

Yes, sorry I didn't clarify that. This is exclusively about the HTTP API. I've been following the protobuf switch and love what you're doing here; I think that's a really smart move. That actually removes an entire element from the documentation as the schema must already be known by all parties ;) I had not considered gRPC as a natural outgrowth... Both this and using protobuf throughout may open up some further avenues. You're spot on regarding client generation, but I'm not sure on the docs side. A quick google search isn't showing anything well-maintained, but conceptually generating docs from `.proto` files + comments would move the docs close to the source and enable relatively straightforward doc generation.

**Comment by ianthetechie** [comment]
Been a while since I've refreshed this thread... We (still) maintain our own handcrafted OpenAPI spec that I'm pretty confident covers ~98% of the HTTP API + data models. In fact, you could probably argue it's _more_ complete than the existing docs, since it fully covers the OSRM response format as well.

My thinking on docs + the state of the art have also shifted since the original issue was opened, so dumping my thoughts here in case they are useful.

# Auto-generating the spec

This is hard. Really hard. It's the way OpenAPI was originally envisioned, and it's definitely _possible_ in some languages with good tooling around the standard, but it's _hard_ regardless. In some cases, there are multiple ways to express something, but you have to pick one to satisfy a particular platform. Quality of generators by language varies _wildly_.

# Spec as docs

The classic OpenAPI generator looks like crap, but it works. Mostly. Until you have really long descriptions and nested data structures and care about formatting so it looks like it's not from 2004.

There are two projects that I'm tracking right now that significantly improve things: [Redoc](https://github.com/Redocly/redoc) and [Scalar](https://github.com/scalar/scalar). Both backed by companies with hosted products. Redoc is significantly more mature at the moment. I've experimented with switching our API docs to Redoc but am hung up on something irrelevant for Valhalla's docs (no ability to edit + send a request directly from the browser).

So _could_ we do this? Probably yes. Do we _want_ to? There are a lot of drawbacks. For one, it reduces searchability. All of these tools rely on JavaScript to dynamically render from an API spec. This makes sense, but is extremely annoying. We have a statically generated site here after all. And we have search too. But that all goes out the window with a dynamically generated doc page. So... meh. (These tools include their own search feature, which of course means you now have TWO search boxes; terrible UX).

I also haven't yet found a doc renderer that is able to show only a subset of a spec. So we'd need to accept a clear divide between API _reference_ and the rest of the docs. I think this is OK actually.

TL;DR there are a lot of challenges. But I have some hope that it will be possible to get a nicer docs-from-spec experience in the future, and IMO maintaining the spec by hand isn't so bad (in fact, I'd volunteer to help with that, since we offer hosted APIs that rely on having correct models and acceptable docs). We're also exploring the same things for our own docs site, so I'll report back here if we make a breakthrough that'd be applicable to the Valhalla docs; ours are very similar (just using MkDocs instead).

**Comment by frankhommers** [comment]
@ianthetechie Would you be able to opensource that self maintained OpenAPI spec? I'd love to use it.

**Comment by ianthetechie** [comment]
Oh cool! Glad to see this finally through @nilsnolde!

@frankhommers yeah it's always been open over here: https://api.stadiamaps.com/openapi.yaml.
